### PR TITLE
Bug: fix issue with nfft<nt

### PR DESCRIPTION
**Pull request by mrava87**
This `PR` handles https://github.com/PyLops/pylops/issues/295 for all FFT operators.

Two changes are made:

1. A warning is raised when user selects `nfft<nt`;
2. When the above condition arises, the output of the adjoint is padded after ifft such as it has size `nt`. This way a user can select `nfft<nt` without getting an error when the adjoint is invoked

### CODE REVIEW

**Comment by mrava87** [comment]
@cako, can you take a look at this.

It was quite easy to handle, as we discussed I both added a warning as most likely users will choose nfft<nt by mistake more than for a real reason. But, if someone has a real reason, now the adjoint works by padding the output of ifft to nt.

File: pylops/signalprocessing/_BaseFFTs.py [hunk 1]
```diff
@@ -102,6 +102,22 @@ def __init__(
             nfft = nffts[0]
         self.nfft = nfft
 
+        # Check if the user provided nnft smaller than n (size of signal in
```
Comment (cako) [inline pylops/signalprocessing/_BaseFFTs.py:105]: Could you fix the typo in `nfft`?

Comment (mrava87) [reply pylops/signalprocessing/_BaseFFTs.py]: Done

**Comment by cako** [comment]
Looks good, I was just working on a patch myself and came across very similar solutions. I'm still reviewing the files but so far it looks good. One thing: my tests flagged this line as problematic: https://github.com/PyLops/pylops/blob/2ac336e9a9b5b40456aecbfcc7e976dbcfdcb1b5/pylops/signalprocessing/FFT.py#L315

when using engine="fftw" and real=True. It might be a good idea to add that test in your suite (it should fail). My fix to this issue was:
```python
np.copyto(self.x, np.take(x, range(self.nfft), axis=self.dir))
```

**Comment by mrava87** [comment]
> Looks good, I was just working on a patch myself and came across very similar solutions. I'm still reviewing the files but so far it looks good. One thing: my tests flagged this line as problematic:
> 
> https://github.com/PyLops/pylops/blob/2ac336e9a9b5b40456aecbfcc7e976dbcfdcb1b5/pylops/signalprocessing/FFT.py#L315
> 
> when using engine="fftw" and real=True. It might be a good idea to add that test in your suite (it should fail). My fix to this issue was:
> 
> ```python
> np.copyto(self.x, np.take(x, range(self.nfft), axis=self.dir))
> ```

Sure. On which test, the `test_FFT_1dsignal`? If I do 

```
par7 = {
    "nt": 41,
    "nx": 31,
    "ny": 10,
    "nfft": 16,
    "real": True,
    "engine": "fftw",
    "ifftshift_before": False,
    "dtype": np.complex128,
}  # nfft<nt, real input, fftw engine
```

it seems to be still fine with the old np.copy. What am I missing?

**Comment by cako** [comment]
Probably nothing hahahaha maybe your implementation is more robust that mine. Just to make sure we are hitting all the right combinations, can you try this test:

```python
par_lists_small_nfft = dict(
    shape=[
        np.random.randint(4, 9, size=(1,)),
        # np.random.randint(4, 9, size=(2,)),
        # np.random.randint(3, 6, size=(3,)),
    ],
    real=[True, False],
    engine=["numpy", "scipy", "fftw"],
)
# Generate all combinations of the above parameters
pars_small_nfft = [
    dict(zip(par_lists_small_nfft.keys(), value))
    for value in itertools.product(*par_lists_small_nfft.values())
]


@pytest.mark.parametrize("par", pars_small_nfft)
def test_small_nfft(par):
    """Dot-test when nfft < nt"""
    shape = par["shape"]
    real = par["real"]
    engine = par["engine"]
    x = np.random.randn(*shape)
    if not real:
        x = x.astype(np.complex128) + 1j * np.random.randn(*shape)

    nffts = [(n + 1) // 2 for n in shape]
    if len(shape) == 1:
        FFTOp = FFT(x.shape, nfft=nffts[0], real=real, engine=engine)
    elif len(shape) == 2:
        if engine == "fftw":
            return
        FFTOp = FFT2D(x.shape, nffts=nffts, real=real, engine=engine)
    else:
        if engine == "fftw":
            return
        FFTOp = FFTND(x.shape, nffts=nffts, real=real, engine=engine)

    y = FFTOp * x.ravel()
    y = y.reshape(FFTOp.dims_fft)
    print("New")
    print("x:     ", x.shape)
    print("nffts: ", nffts)
    print("y:     ", y.shape)
    xadj = FFTOp.H * y.ravel()
    xadj = xadj.reshape(x.shape)
    print("xadj:  ", xadj.shape)
    nr, nc = FFTOp.shape
    if par["real"]:
        assert dottest(FFTOp, nr, nc, complexflag=2)
    else:
        assert dottest(FFTOp, nr, nc, complexflag=2)
        assert dottest(FFTOp, nr, nc, complexflag=3)

```

**Comment by mrava87** [comment]
Mmmh I just copy-pasted this at the bottom of the `test_ffts.py` file and all tests pass :)

File: pylops/signalprocessing/FFT.py [hunk 2]
```diff
@@ -185,8 +189,12 @@ def _rmatvec(self, x):
             y = scipy.fft.ifft(x, n=self.nfft, axis=self.dir, **self._norm_kwargs)
         if self.norm is _FFTNorms.NONE:
             y *= self._scale
-        if self.nfft != self.dims[self.dir]:
+
+        if self.nfft > self.dims[self.dir]:
             y = np.take(y, np.arange(0, self.dims[self.dir]), axis=self.dir)
```
Comment (cako) [inline pylops/signalprocessing/FFT.py:194]: I think this is a good opportunity to change all of these lines to the simpler:
```python
np.take(y, range(self.dims[self.dir]), axis=self.dir)
```

File: pylops/signalprocessing/FFT2D.py [hunk 3]
```diff
@@ -93,8 +93,14 @@ def _rmatvec(self, x):
...
             y *= self._scale
-        y = np.take(y, range(self.dims[self.dirs[0]]), axis=self.dirs[0])
-        y = np.take(y, range(self.dims[self.dirs[1]]), axis=self.dirs[1])
+        if self.nffts[0] > self.dims[self.dirs[0]]:
+            y = np.take(y, range(self.dims[self.dirs[0]]), axis=self.dirs[0])
+        if self.nffts[1] > self.dims[self.dirs[1]]:
+            y = np.take(y, range(self.dims[self.dirs[1]]), axis=self.dirs[1])
+        if (self.nffts[0] < self.dims[self.dirs[0]]) or (
+            self.nffts[1] < self.dims[self.dirs[1]]
+        ):
```
Comment (cako) [inline pylops/signalprocessing/FFT2D.py:102]: Why note use `self.doifftpad` here?

File: pylops/signalprocessing/FFTND.py [hunk 4]
```diff
@@ -84,8 +84,16 @@ def _rmatvec(self, x):
...
-            y = np.take(y, range(self.dims[direction]), axis=direction)
+        for direction, nfft in zip(self.dirs, self.nffts):
+            if nfft > self.dims[direction]:
+                y = np.take(y, range(self.dims[direction]), axis=direction)
+        if any(
+            [
+                nfft < self.dims[direction]
+                for direction, nfft in zip(self.dirs, self.nffts)
+            ]
+        ):
```
Comment (cako) [inline pylops/signalprocessing/FFTND.py:95]: Why note use `self.doifftpad` here?

File: pylops/signalprocessing/FFTND.py [hunk 5]
```diff
@@ -169,8 +177,16 @@ def _rmatvec(self, x):
...
```
Comment (cako) [inline pylops/signalprocessing/FFTND.py:188]: And here

File: pylops/signalprocessing/FFT.py (hunk 2 above)
Comment (mrava87) [reply pylops/signalprocessing/FFT.py:194]: Agreed!

File: pylops/signalprocessing/FFT2D.py (hunk 3 above)
Comment (mrava87) [reply pylops/signalprocessing/FFT2D.py:102]: Good one :)

File: pylops/signalprocessing/FFTND.py (hunk 5 above)
Comment (mrava87) [reply pylops/signalprocessing/FFTND.py:188]: Same

**Comment by mrava87** [comment]
Added all! Let me know if you agree and I merge it :)

**Review by cako — APPROVED** [review]
I'm approving but there are some changes not related to the bug. Since they are comments I don't think it's a big issue.

**Pull request merged**